### PR TITLE
Fix outputs wrongly evaluating as optional [BT-281]

### PR DESCRIPTION
**Pull request by mcovarr**

### CODE REVIEW

**Review by kshakir — APPROVED** [review]
Very generic LGTM. I'm not familiar with how this plugs into the wider application.

ToL: In a dream world a previous set of expectation unit tests would have been updated by this behavioral change.

But I don't see any evidence of a `WomExpressionSpec` and don't know the area enough to know if there's another Spec structurally testing `areAllFileTypesInWomTypeOptional`.

BT-281 didn't go into too much detail so to be explicit I'm also assuming the added centaur WDL was broken before this PR. If what existing tests are there continue to pass then the new integration test acts as a nice regression test. 👍

**Comment by mcovarr** [comment]
Yes the new test failed previously, the output would have been marked as `optional` when it should have been `required`. Existing tests for cases where file outputs should be optional are still passing, so yay.

File: centaur/src/main/resources/standardTestCases/papi_delocalization_required_files/papi_delocalization_required_files.wdl [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        String value = read_string("file.txt")
+    }
+    runtime {
+        docker: "ubuntu:latest"
+    }
+}
+
+workflow required_files {
+    call check_it
+    call do_it { input: sync = check_it.value }
```
Comment (cjllanwarne) [inline centaur/src/main/resources/standardTestCases/papi_delocalization_required_files/papi_delocalization_required_files.wdl:39]: in WDL 1.1 (or `version development`) you can use `call do_it after check_it`

**Review by cjllanwarne — COMMENTED** [review]
It seems like the problem here is the `read_string` requiring the file to be downloaded, more than the type definition of its output.

Eg wouldn't you still have the same problem if you changed the output from `String value = read_string("file.txt")` to `Array[String] value = [ read_string("file.txt") ]` ?

**Comment by mcovarr** [comment]
@cjllanwarne maybe? We may be getting into areas where I'm not familiar with the specification (or it may not be specifying). Should this should make `file.txt` optional?

```
String? value = read_string("file.txt")
```

And if so could an array be thought of as a generalization of an optional from at most one to arbitrarily many instances?

At any rate these changes were focused on having

```
String value = read_string("file.txt")
```

make `file.txt` required, which I believe is more correct than the current behavior.

**Comment by cjllanwarne** [comment]
In `String? value = read_string("file.txt")` the `file.txt` is non-optional, the `read_string` will always evaluate, and the `String? value` will always be filled in (via coercion from `String` to `String?`.

**Comment by mcovarr** [comment]
@cjllanwarne any suggestions for alternative changes to make here? Or is this PR just not a good idea?

**Comment by cjllanwarne** [comment]
I think it's fine, it's no _more_ broken than before and unblocks a previously unworking case. It moves the ball forward, it's just not a panacea

File: centaur/src/main/resources/standardTestCases/papi_delocalization_required_files/papi_delocalization_required_files.wdl (hunk 1 above)
Comment (breilly2) [reply centaur/src/main/resources/standardTestCases/papi_delocalization_required_files/papi_delocalization_required_files.wdl:39]: I'm having nightmare flashbacks of reading perl...
```
do_something() if allowed
```

File: wom/src/main/scala/wom/expression/WomExpression.scala [hunk 2]
```diff
@@ -23,23 +23,35 @@ trait WomExpression {
...
+
   def inputs: Set[String]
+
   def evaluateValue(inputValues: Map[String, WomValue], ioFunctionSet: IoFunctionSet): ErrorOr[WomValue]
+
   def evaluateType(inputTypes: Map[String, WomType]): ErrorOr[WomType]
+
   def evaluateFiles(inputValues: Map[String, WomValue], ioFunctionSet: IoFunctionSet, coerceTo: WomType): ErrorOr[Set[FileEvaluation]]
 
   /** Returns `true` if all file types within the specified `WomType` are optional. If not all the file types are
```
Comment (breilly2) [inline wom/src/main/scala/wom/expression/WomExpression.scala:35]: Is this doc still correct? I guess I don't understand why a primitive type would be passed to this function based on these docs.

**Review by breilly2 — APPROVED** [review]
👍   Curious about the comment/code, but with the added test, this seems to do what you want it to do.

File: wom/src/main/scala/wom/expression/WomExpression.scala (hunk 2 above)
Comment (mcovarr) [reply wom/src/main/scala/wom/expression/WomExpression.scala:35]: Yes this is still correct, it's looking at the type of the output variable which could be a primitive.

**Pull request merged**